readme.txt
=== WooCommerce Gravity Forms Product Add-ons ===
Contributors: lucasstark
Tags: woocommerce, gravity forms
Requires at least: 3.0
Tested up to: 4.9.5

WooCommerce Gravity Forms.  This add-on requires Gravity Forms and has been tested with Gravity Forms 2.3

== Installation ==

1. Upload the folder 'woocommerce-gravityforms-product-addons' to the '/wp-content/plugins/' directory

2. Activate 'WooCommerce - Gravity Forms Product Add-Ons' through the 'Plugins' menu in WordPress

== Usage ==

1. Edit a product and look for the 'Gravity Forms' write panel.  The product must have a non-empty price, enter 0.00 if the product is free. 

2. Choose a gravity form that you would like to use as a product form for your product. 

3. Enter in any text before / after the standard price.  This allows you to add something such as Starts:  
for products that are not variable, but based on the gravity form options will have a variable price. 

4: Update / publish the product

5: Browse to the product from your store, fill out the gravity form, and validate that all the correct information is being collected 
throughout the process. 

-- Gravity Forms -- 
Build a gravity form using any standard form field you require. 

This add-on supports both standard data collection fields ( such as a text box ), and pricing fields that effect the price.  If you would like to use pricing fields on your form, build a Gravity Form with a Product Field and a Total field. Typically you will set the Product Field to be hidden and set the price to 0.  You would then use additional pricing fields to modify the price of the item as options are selected.    If you are building a donation style form, you can use the product field and allow a user defined price.  

If you are building a Gravity Form that includes pricing fields, you can add the css class of hidden-total to the total field on the gravity form to hide this line item.  WooCommerce-Gravity Forms includes a built in function that will display the total of the customizations selected on the gravity form when a user is adding the item to their cart. 

-- WooCommerce --
WooCommerce 2.4 or greater is required, WooCommerce 2.7 or greater is recommended.

Grouped products are not supported.

On the product admin screen you will see a Gravity Forms meta box.  This meta box allows you to choose the gravity form to be linked to the product as well as configure various display options for the form.

changelog.txt
*** Gravity Forms Product Add-Ons Changelog ***
2020.08.18 - version 3.3.15
    * New:  Added filter, woocommerce_gforms_script_params, to allow modification of the script params object enqueued for the forms front end.
    * Update:  Added additional logging when an order is being created.

2020.06.25 - version 3.3.14
    * Update:  WC and WP version compatibility tags.
    * Fix:  Fix issue with email fields which require email confirmation on subscription renewal.
    * Fix:  Short circuit get_item_data if subscription_initial_payment is present on the cart item because subscriptions stores the meta data and automatically adds it back in.
                If we allow this to run, we end up with duplicate cart item meta data.

2020.05.06 - version 3.3.12
    * Fix:  Make sure variation exists before attempting to get the price.
    * Update:  WC and WP version compatibility tags.

2020.04.15 - version 3.3.11
    * Update:  WC and WP Version Bump.

2020.02.19 - version 3.3.10
    * Update:  Allow for the CSS class of wc-gforms-hide-from-email to work on Product Fields in the cart.
    * Update:  Allow for the CSS class of wc-gforms-hide-from-email to work on Product Fields during checkout and the order process.
    * Update:  WC and WP Version Bump.

2019.09.16 - version 3.3.9
    * Update:  Include Structured Data Helper
            Helpful when the base price of the product is 0, or when you want Google to be aware of the pricing options in your form.
    * Update:  Added translations for some more admin options which were missing them.
    * Update:  Hide add to cart buttons via CSS when on a multipage form and not on the last page.
    * Update:  Update the gform entry metadata before sending notifications so that the WC Order information can be used.
    * Update:  Fields marked with the Visibility: Hidden attribute will be hidden in the WC Order details.
    * Update:  Update Dynamic Pricing calculation so that variable product rules can be taken into account.
    * Update:  Allow for drop down quantity fields.
    * Update:  Make sure fields with a "0" value are included in the cart, orders, and confirmation emails.
    * Update:  Fix display issue with custom hidden fields in the Order display screen.

2019.02.27 - version 3.3.8
    * Update:  Include new filter woocommerce_gforms_order_item_meta to allow filtering of order item meta data just prior to it being added to the WC order item.
    * Update:  Perform a check before adding order item meta data to prevent duplicates.  This could only be caused if the filter was being called incorrectly twice for some reason.
    * Update:  Include new filter woocommerce_gforms_get_cart_item_total to filter the gforms total to be added to the cart item.
    * Update:  Include the ability to attach a gravity form to a products bulk input form when WooCommerce Bulk Variation Forms are in use.

2019.01.02 - version 3.3.7
    * Update:  Tested to WC 3.5.3
    * Update:  Tested to WP 5.0.2
    * Update:  Fix issue where 0 values were not being displayed in the cart.
    * Fix:  Fix for validation errors when the user fails to fill out all the required fields on a form more than once.
    * New:  Add option to display totals section before the form.  Helpful for multi-page forms.
    * New:  Include separate Bulk Variations Form selection option.  Fixes issues with duplicate forms being rendered.
    * Fix:  Fix issue where the setting for Replace Modified Items was not being saved.
    * Fix:  Fix issue with the gform_product_total and not respecting the Use Dynamic Pricing filter.
    * Fix:  Fix warning in PHP 7.3 about continue inside of a switch statement.

2018.10.23 - version 3.3.6
    * Update:  Use basename() for file names rather than attempting to parse ourselves.
    * Update:  Add hidden class to the metaboxes to help fix up some bouncing in the UI when editing products.
    * Update:  Add additional arguments to the woocommerce_order_item_display_meta_value call on the order screen.
    * Update:  Tested to WC 3.5

2018.08.22 - version 3.3.5
    * Update:  Only use file names when creating the order item meta.
    * Fix:  Fix issue with the reorder validation.
    
2018.08.09 - version 3.3.4
    * Fix:  Fix calculation when using Dynamic Pricing.
    * Update:  WP version update.
    * Update:  Make sure the same form is attached to the product when re-ordering, otherwise re-order is disallowed.
    * Update:  Fill in email confirmation field when using the cart edit feature.
    * Update:  Use products regular base price when performing AJAX calculations for Dynamic Pricing.

2018.06.29 - version 3.3.3
    * Fix:  Update created_by property instead of user on the entries created during checkout.
    * Update:  Remove - from plugin name.
    * Update:  WC and WP version updates.

2018.06.05 - version 3.3.2
    * New:  Add woocommerce_gravityforms_wcs_new_order_created_create_entries to allow disabling of new entry creation on subscription renewal.
    * Update:  Prevent WooCommerce Subscriptions from causing duplicate entry information.
    * Update:  Fire validation for woocommerce_order_again_cart_item_data.
    * Fix:  Account for checkbox product option fields when using the cart edit option.
    * Fix:  Fix Cart Edit - Replace Modified Items.
    * Update:  Include class GFFormDisplay in get_item_data in case developers use gravity form hooks which expect that to be loaded.
    * Update:  Allow forms to work on external products.

2018.05.11 - version 3.3.1
    * Fix:  Fix warning about missing function in the entry management class.
    * Fix:  Fix issue where we did not return data from on_get_order_again_cart_item_data.

2018.04.26 - version 3.3.0
    * Note:   Tested up to Gravity Forms 2.3
    * Fix:    Entry deletion for Gravity Forms 2.3
    * Fix:    Scroll to form automatically if there are validation errors.
    * Update: Include Order Status column as part of the Gravity Forms Entry export feature.

compatibility.php
<?php

/**
 * WooCommerce Plugin Compatibility
 *
 * This source file is subject to the GNU General Public License v3.0
 * that is bundled with this package in the file license.txt.
 * It is also available through the world-wide-web at this URL:
 * http://www.gnu.org/licenses/gpl-3.0.html
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@skyverge.com so we can send you a copy immediately.
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade the plugin to newer
 * versions in the future. If you wish to customize the plugin for your
 * needs please refer to http://www.skyverge.com
 *
 * @author    SkyVerge
 * @copyright Copyright (c) 2013-2014, SkyVerge, Inc.
 * @license   http://www.gnu.org/licenses/gpl-3.0.html GNU General Public License v3.0
 */
if ( !defined( 'ABSPATH' ) )
	exit; // Exit if accessed directly

if ( !class_exists( 'WC_GFPA_Compatibility' ) ) :

	/**
	 * WooCommerce Compatibility Utility Class
	 *
	 * The unfortunate purpose of this class is to provide a single point of
	 * compatibility functions for dealing with supporting multiple versions
	 * of WooCommerce.
	 *
	 * The recommended procedure is to rename this file/class, replacing "my plugin"
	 * with the particular plugin name, so as to avoid clashes between plugins.
	 * Over time we expect to remove methods from this class, using the current
	 * ones directly, as support for older versions of WooCommerce is dropped.
	 *
	 * Current Compatibility: 2.1.x - 2.2
	 *
	 * @version 2.0
	 */
	class WC_GFPA_Compatibility {

		/**
		 * Get the WC Order instance for a given order ID or order post
		 *
		 * Introduced in WC 2.2 as part of the Order Factory so the 2.1 version is
		 * not an exact replacement.
		 *
		 * If no param is passed, it will use the global post. Otherwise pass an
		 * the order post ID or post object.
		 *
		 * @since 2.0.0
		 * @param bool|int|string|\WP_Post $the_order
		 * @return bool|\WC_Order
		 */
		public static function wc_get_order( $the_order = false ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return wc_get_order( $the_order );
			} else {

				global $post;

				if ( false === $the_order ) {

					$order_id = $post->ID;
				} elseif ( $the_order instanceof WP_Post ) {

					$order_id = $the_order->ID;
				} elseif ( is_numeric( $the_order ) ) {

					$order_id = $the_order;
				}

				return new WC_Order( $order_id );
			}
		}

		/**
		 * Get all order statuses
		 *
		 * Introduced in WC 2.2
		 *
		 * @since 2.0.0
		 * @return array
		 */
		public static function wc_get_order_statuses() {

			if ( self::is_wc_version_gte_2_2() ) {

				return wc_get_order_statuses();
			} else {

				// get available order statuses
				$order_status_terms = get_terms( 'shop_order_status', array('hide_empty' => false) );

				if ( is_wp_error( $order_status_terms ) ) {

					$order_status_terms = array();
				}

				$order_statuses = array();

				foreach ( $order_status_terms as $term ) {

					$order_statuses[$term->slug] = $term->name;
				}

				return $order_statuses;
			}
		}

		/**
		 * Get the order status
		 *
		 * Order statuses changed from a taxonomy in 2.1 to using `post_status`
		 * in 2.2
		 *
		 * @since 2.0.0
		 * @param WC_Order $order
		 * @return mixed|string
		 */
		public static function get_order_status( WC_Order $order ) {

			return self::is_wc_version_gte_2_2() ? $order->get_status() : $order->status;
		}

		/**
		 * Check if the order has the given status
		 *
		 * @param WC_Order $order
		 * @param string|array $status single status or array of statuses to check
		 * @since 2.0.0
		 * @return bool
		 */
		public static function order_has_status( $order, $status ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return $order->has_status( $status );
			} else {

				if ( is_array( $status ) ) {

					return in_array( $order->status, $status );
				} else {

					return $status === $order->status;
				}
			}
		}

		/**
		 * Transparently backport the `post_status` WP Query arg used by WC 2.2
		 * for order statuses to the `shop_order_status` taxonomy query arg used by
		 * WC 2.1
		 *
		 * @since 2.0.0
		 * @param array $args WP_Query args
		 * @return array
		 */
		public static function backport_order_status_query_args( $args ) {

			if ( !self::is_wc_version_gte_2_2() ) {

				// convert post status arg to taxonomy query compatible with WC 2.1
				if ( !empty( $args['post_status'] ) ) {

					$order_statuses = array();

					foreach ( (array) $args['post_status'] as $order_status ) {

						$order_statuses[] = str_replace( 'wc-', '', $order_status );
					}

					$args['post_status'] = 'publish';

					$tax_query = array(
					    array(
						'taxonomy' => 'shop_order_status',
						'field' => 'slug',
						'terms' => $order_statuses,
						'operator' => 'IN',
					    )
					);

					$args['tax_query'] = array_merge( ( isset( $args['tax_query'] ) ? $args['tax_query'] : array() ), $tax_query );
				}
			}

			return $args;
		}

		/**
		 * Get the user ID for an order
		 *
		 * @since 2.0.0
		 * @param \WC_Order $order
		 * @return int
		 */
		public static function get_order_user_id( WC_Order $order ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return $order->get_user_id();
			} else {

				return $order->customer_user ? $order->customer_user : 0;
			}
		}

		/**
		 * Get the user for an order
		 *
		 * @since 2.0.0
		 * @param \WC_Order $order
		 * @return bool|WP_User
		 */
		public static function get_order_user( WC_Order $order ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return $order->get_user();
			} else {

				return self::get_order_user_id( $order ) ? get_user_by( 'id', self::get_order_user_id( $order ) ) : false;
			}
		}

		/**
		 * Get the WC Product instance for a given product ID or post
		 *
		 * get_product() is soft-deprecated in WC 2.2
		 *
		 * @since 2.0.0
		 * @param bool|int|string|\WP_Post $the_product
		 * @param array $args
		 * @return WC_Product
		 */
		public static function wc_get_product( $the_product = false, $args = array() ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return wc_get_product( $the_product, $args );
			} else {

				return get_product( $the_product, $args );
			}
		}

		/**
		 * Return an array of formatted item meta in format:
		 *
		 * array(
		 *   $meta_key => array(
		 *     'label' => $label,
		 *     'value' => $value
		 *   )
		 * )
		 *
		 * e.g.
		 *
		 * array(
		 *   'pa_size' => array(
		 *     'label' => 'Size',
		 *     'value' => 'Medium',
		 *   )
		 * )
		 *
		 * Backports the get_formatted() method to WC 2.1
		 *
		 * @since 2.2.0
		 * @see WC_Order_Item_Meta::get_formatted()
		 * @param \WC_Order_Item_Meta $item_meta order item meta class instance
		 * @param string $hide_prefix exclude meta when key is prefixed with this, defaults to `_`
		 * @return array
		 */
		public static function get_formatted_item_meta( WC_Order_Item_Meta $item_meta, $hide_prefix = '_' ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return $item_meta->get_formatted( $hide_prefix );
			} else {

				if ( empty( $item_meta->meta ) ) {
					return array();
				}

				$formatted_meta = array();

				foreach ( (array) $item_meta->meta as $meta_key => $meta_values ) {

					if ( empty( $meta_values ) || !is_array( $meta_values ) || (!empty( $hide_prefix ) && substr( $meta_key, 0, 1 ) == $hide_prefix ) ) {
						continue;
					}

					foreach ( $meta_values as $meta_value ) {

						// Skip serialised meta
						if ( is_serialized( $meta_value ) ) {
							continue;
						}

						$attribute_key = urldecode( str_replace( 'attribute_', '', $meta_key ) );

						// If this is a term slug, get the term's nice name
						if ( taxonomy_exists( $attribute_key ) ) {
							$term = get_term_by( 'slug', $meta_value, $attribute_key );

							if ( !is_wp_error( $term ) && is_object( $term ) && $term->name ) {
								$meta_value = $term->name;
							}

							// If we have a product, and its not a term, try to find its non-sanitized name
						} elseif ( $item_meta->product ) {
							$product_attributes = $item_meta->product->get_attributes();

							if ( isset( $product_attributes[$attribute_key] ) ) {
								$meta_key = wc_attribute_label( $product_attributes[$attribute_key]['name'] );
							}
						}

						$formatted_meta[$meta_key] = array(
						    'label' => wc_attribute_label( $attribute_key ),
						    'value' => apply_filters( 'woocommerce_order_item_display_meta_value', $meta_value ),
						);
					}
				}

				return $formatted_meta;
			}
		}

		/**
		 * Get the full path to the log file for a given $handle
		 *
		 * @since 2.0.0
		 * @param string $handle log handle
		 * @return string
		 */
		public static function wc_get_log_file_path( $handle ) {

			if ( self::is_wc_version_gte_2_2() ) {

				return wc_get_log_file_path( $handle );
			} else {

				return sprintf( '%s/plugins/woocommerce/logs/%s-%s.txt', WP_CONTENT_DIR, $handle, sanitize_file_name( wp_hash( $handle ) ) );
			}
		}

		/**
		 * Gets a product meta field value, regardless of product type
		 *
		 * @since 2.0
		 * @param WC_Product $product the product
		 * @param string $field_name the field name
		 * @return mixed meta value
		 */
		public static function get_product_meta( $product, $field_name ) {

			// Should be able to drop this when we drop support for WC 2.1
			// even in WC >= 2.0 product variations still use the product_custom_fields array apparently
			if ( $product->variation_id && isset( $product->product_custom_fields['_' . $field_name][0] ) && '' !== $product->product_custom_fields['_' . $field_name][0] ) {
				return $product->product_custom_fields['_' . $field_name][0];
			}

			// use magic __get
			return $product->$field_name;
		}

		public static function wc_error_count() {
			if ( self::is_wc_version_gt( '2.0' ) ) {
				return wc_notice_count( 'error' );
			} else {
				global $woocommerce;
				return $woocommerce->error_count();
			}
		}

		/**
		 * Helper method to get the version of the currently installed WooCommerce
		 *
		 * @since 2.0.0
		 * @return string woocommerce version number or null
		 */
		private static function get_wc_version() {
			return defined( 'WC_VERSION' ) && WC_VERSION ? WC_VERSION : WC()->version;
		}

		/**
		 * Returns true if the installed version of WooCommerce is 2.2 or greater
		 *
		 * @since 2.2.0
		 * @return boolean true if the installed version of WooCommerce is 2.2 or greater
		 */
		public static function is_wc_version_gte_2_2() {
			return self::get_wc_version() && version_compare( self::get_wc_version(), '2.2', '>=' );
		}
		
		/**
		 * Returns true if the installed version of WooCommerce is 2.5 or greater
		 *
		 * @since 2.4.0
		 * @return boolean true if the installed version of WooCommerce is 2.5 or greater
		 */
		public static function is_wc_version_gte_2_4() {
			return self::get_wc_version() && version_compare( self::get_wc_version(), '2.4', '>=' );
		}
		
		/**
		 * Returns true if the installed version of WooCommerce is 2.5 or greater
		 *
		 * @since 2.5.0
		 * @return boolean true if the installed version of WooCommerce is 2.5 or greater
		 */
		public static function is_wc_version_gte_2_5() {
			return self::get_wc_version() && version_compare( self::get_wc_version(), '2.5', '>=' );
		}

		/**
		 * Returns true if the installed version of WooCommerce is 2.7 or greater
		 *
		 * @since 3.0.0
		 * @return boolean true if the installed version of WooCommerce is 2.7 or greater
		 */
		public static function is_wc_version_gte_2_7() {
			return self::get_wc_version() && version_compare( self::get_wc_version(), '2.7', '>=' );
		}

		/**
		 * Returns true if the installed version of WooCommerce is less than 2.2
		 *
		 * @since 2.2.0
		 * @return boolean true if the installed version of WooCommerce is less than 2.2
		 */
		public static function is_wc_version_lt_2_2() {
			return self::get_wc_version() && version_compare( self::get_wc_version(), '2.2', '<' );
		}

		/**
		 * Returns true if the installed version of WooCommerce is 2.7 or greater
		 *
		 * @since 3.0.0
		 * @return boolean true if the installed version of WooCommerce is 2.7 or greater
		 */
		public static function is_wc_version_gte_3_2() {
			return self::get_wc_version() && version_compare( self::get_wc_version(), '3.2', '>=' );
		}

		/**
		 * Returns true if the installed version of WooCommerce is greater than $version
		 *
		 * @since 2.0.0
		 * @param string $version the version to compare
		 * @return boolean true if the installed version of WooCommerce is > $version
		 */
		public static function is_wc_version_gt( $version ) {
			return self::get_wc_version() && version_compare( self::get_wc_version(), $version, '>' );
		}

	}

	endif; // Class exists check



if ( !function_exists( 'wc_is_21x' ) ) {

	function wc_is_21x() {
		return function_exists( 'wc_add_notice' );
	}

}

if ( !function_exists( 'is_wc_version_gte_2_2' ) ) {

	/**
	 * Returns true if the installed version of WooCommerce is 2.2 or greater
	 *
	 * @since 1.0
	 * @return boolean true if the installed version of WooCommerce is 2.1 or greater
	 */
	function is_wc_version_gte_2_2() {
		global $woocommerce;
		return version_compare( $woocommerce->version, '2.2.0', 'ge' );
	}

}

if ( !function_exists( 'is_wc_version_gte_2_3' ) ) {

	/**
	 * Returns true if the installed version of WooCommerce is 2.3 or greater
	 *
	 * @since 1.0
	 * @return boolean true if the installed version of WooCommerce is 2.1 or greater
	 */
	function is_wc_version_gte_2_3() {
		global $woocommerce;
		return version_compare( $woocommerce->version, '2.3.0', 'ge' );
	}

}

if ( !function_exists( 'is_wc_version_gte_2_3' ) ) {

	/**
	 * Returns true if the installed version of WooCommerce is 2.3 or greater
	 *
	 * @since 1.0
	 * @return boolean true if the installed version of WooCommerce is 2.1 or greater
	 */
	function is_wc_version_gte_2_3() {
		global $woocommerce;

		return version_compare( $woocommerce->version, '2.3.0', 'ge' );
	}

}

